### DISCUSSION TITLE
Importing named exports from CJS modules works when deps installed with npm but not with yarn?

### DISCUSSION THREAD

**Discussion by achingbrain** [webpack/webpack#14268] [Q&A]
# Bug report

**What is the current behavior?**

Given `a-cjs-module` where the index.js file contains something like:

```js
const someValue = () => {}

module.exports = { someValue }
```

Running this code on node works, treating the properties of `module.exports` as ESM named exports:

```js
import { someValue } from 'a-cjs-module'
```

See:

> Named exports may be available, provided by static analysis as a convenience for better ecosystem compatibility

https://nodejs.org/api/esm.html#esm_interoperability_with_commonjs

Running it on webpack fails:

```
../../node_modules/some-module/index.js:83:41-47 - Error: export 'someValue' (imported as 'someValue') was not found in 'a-cjs-module' (possible exports: __esModule, somethingExportedFromIndexJs)
```

See:

> Only the "default" export can be imported from non-ESM. Named exports are not available.

https://webpack.js.org/guides/ecma-script-modules/#flagging-modules-as-esm

**If the current behavior is a bug, please provide the steps to reproduce.**

See the code samples above.

**What is the expected behavior?**

I would expect webpack to perform the same static analysis node does and make the values available as a named exports for better ecosystem compatibility.

Many modules have followed the `module.exports = { someValue }` pattern to be better compatible with this bright named-exports-based future and to ease upgrading when the time comes to fully switch from CJS to ESM, but now none of them work in webpack.

**Other relevant information:**
webpack version: 5.53.0
Node.js version: 16.8.0
Operating System: Mac OS 
Additional tools: N/a

**Comment by alexander-akait** [comment]
It was implemented a long time ago and no problems for me, 

note:
```
../../node_modules/some-module/index.js:83:41-47 - Error: export 'someValue' (imported as 'someValue') was not found in 'a-cjs-module' (possible exports: __esModule, somethingExportedFromIndexJs)
```

You have named export, just another name - `somethingExportedFromIndexJs

**Reply by achingbrain** [reply]
> You have named export, just another name

The error message is misleading - `importer` is defined by the importing module (e.g. `ipfs-unixfs-importer`), not the exporting module (e.g. `rabin-wasm`).  The actual message is this:

```
ERROR in ./node_modules/ipfs-unixfs-importer/esm/src/chunker/rabin.js 39:18-24
export 'create' (imported as 'create') was not found in 'rabin-wasm' (possible exports: __esModule, importer)
```

To replicate:

```console
$ git clone git@github.com:achingbrain/webpack-cjs-esm-yarn-npm.git
$ cd webpack-cjs-esm-yarn-npm
$ yarn
$ yarn run build

... other output

ERROR in ./node_modules/ipfs-unixfs-importer/esm/src/chunker/rabin.js 39:18-24
export 'create' (imported as 'create') was not found in 'rabin-wasm' (possible exports: __esModule, importer)
 @ ./node_modules/ipfs-unixfs-importer/esm/src/dag-builder/index.js 4:0-40 44:18-23
 @ ./node_modules/ipfs-unixfs-importer/esm/src/index.js 3:0-50 11:17-29
 @ ./node_modules/ipfs-core/src/components/add-all/index.js 3:21-52
 @ ./node_modules/ipfs-core/src/components/root.js 4:24-44
 @ ./node_modules/ipfs-core/src/components/index.js 30:16-33
 @ ./node_modules/ipfs-core/src/index.js 10:19-42
 @ ./node_modules/ipfs/src/index.js 7:22-42
 @ ./src/index.ts 12:15-30

... lots of similar errors
```

* `ipfs-unixfs-importer` is an ESM module that's dual-published as ESM and CJS.  It is depended on by the project being built by webpack.
* [ipfs-unixfs-importer/src/index.js](https://unpkg.com/ipfs-unixfs-importer@9.0.6/esm/src/index.js) has a named export of `importer`.
* `ipfs-unixfs-importer` depends on `rabin-wasm` which is a CJS module.
* [ipfs-unixfs-importer/esm/src/chunker/rabin.js](https://unpkg.com/ipfs-unixfs-importer@9.0.6/esm/src/chunker/rabin.js) contains:

```
import { create } from 'rabin-wasm';
```

* [rabin-wasm/src/index.js](https://unpkg.com/rabin-wasm@0.1.5/src/index.js) contains:

```
module.exports = {
    create
}
```

Weirdly this all works as expected if you install the deps with npm, it only fails if you've used yarn.

**Comment by achingbrain** [comment]
@alexander-akait I think you closed my issue prematurely, though granted it was a bit inaccurate.

I've done a bit more digging - the key difference between the `node_modules` folders that npm and yarn generate is that yarn creates `node_module` folders in each dep where their deps have cli tools with a nested `.bin` folder and npm does not.

So from the above, `ipfs-unixfs-importer` depends on `rabin-wasm` which has a [cli tool](https://github.com/hugomrdias/rabin-wasm/blob/master/package.json#L11).

With npm I get:

```
my-project/
├── node_modules/
│   ├── .bin/
│   │   └── rabin-wasm
.. other files
│   ├── ipfs-unixfs-importer/
|   |   └── package.json // etc
```

With yarn I get:

```
my-project/
├── node_modules/
│   ├── .bin/
│   │   └── rabin-wasm
.. other files
│   ├── ipfs-unixfs-importer/
│   │   ├── node_modules/
│   │   │   └── .bin/
│   │   │       └── rabin-wasm
|   |   └── package.json // etc
```

If I remove (or rename) the nested `my-project/node_modules/ipfs-unixfs-importer/node_modules` folder everything works, if I leave it, webpack does not resolve the `rabin-wasm` dependency.  I guess it doesn't walk back up the tree to find the next `node_modules` folder if the first one doesn't have the dep it's looking for like node does?

You can see this yourself with the [repro repo](https://github.com/achingbrain/webpack-cjs-esm-yarn-npm).

**Reply by alexander-akait** [reply]
>  I think you closed my issue prematurely, though granted it was a bit inaccurate.

Nobody closed anything, it was translated into a discussion, and as you can see, the original problem does not coincide with what you encountered, that is why it will be investigated here, and in the future an answer will be given or a new issue created in the correct place and description

We don't have any special logic between npm and yarn (except PnP, but you don't use it in repo) and etc, and mostly it is bug in yarn, I will look at this

**Reply by achingbrain** [reply]
Thanks for looking into it.

FWIW I don't think it's specific to yarn, it just happens that yarn generates a node_modules folder structure that triggers the bug.

I'm starting to see it in different projects using npm too but am struggling to get a minimum reproducible case just yet.

**Reply by alexander-akait** [reply]
hm looks like problem in the `exports` field, `enhanced-resolve` try to find `rabin-wasm`, but because `node_modules` exists, `enhanced-resolve` does not go to up and uses the `exports` field, try to remove `exports` from `ipfs-unixfs-importer/package.json` and all works fine, very weird, still investigate
